examples/QIR/Simulation/StateSimulator/README.md

```diff
@@ -0,0 +1,375 @@
+The state of the active qubits in this Hilbert space is stored in the `stateVector` member:
+
+```cpp
+class StateSimulator : public IRuntimeDriver, public IQuantumGateSet
+{
+    // Associated qubit manager instance to handle qubit representation.
+    CQubitManager *qbm;
+
+    // The register of currently active qubits.
+    short numActiveQubits = 0;
+    std::vector<Qubit> computeRegister;
+
+    // The state of the compute register is represented by its full 2^n column vector of probability amplitudes.
+    // With no qubits allocated, the state vector starts out as the scalar 1.
+    State stateVec = State::Ones(1);
+```
+
+The helper functions below deal with updating the state vector for new or deallocated qubits, apply `Gate` or multi-controlled `Gate` operations to the compute register, and build operators over the active qubit space made up of tensor products of pauli matrices.
+
+```cpp
+    // To be called on allocation/deallocation of qubits to update the state vector.
+    void UpdateState(short qubitIndex, bool remove = false);
+
+    // To be called by quantum gate set operations.
+    void ApplyGate(Gate gate, Qubit target);
+    void ApplyControlledGate(Gate gate, long numControls, Qubit controls[], Qubit target);
+
+    // Builds a unitary matrix over the state space made of Pauli operators.
+    Operator BuildPauliUnitary(long numTargets, PauliId paulis[], Qubit targets[]);
+```
+
+A new qubit manager instance can simply be attached to the simulator in the constructor, which also initializes the PRNG with a provided seed:
+
+```cpp
+    StateSimulator(uint32_t userProvidedSeed = 0)
+    {
+        srand(userProvidedSeed);
+        this->qbm = new CQubitManager();
+    }
+    ~StateSimulator()
+    {
+        delete this->qbm;
+    }
+```
+
+The following functions will be used from the QIR qubit manager, but additional functionality is present to manage how qubits are reused (full interface at [public/QubitManager.hpp](https://github.com/microsoft/qsharp-runtime/blob/main/src/Qir/Runtime/public/QubitManager.hpp)):
+
+```cpp
+Qubit CQubitManager::Allocate()
+CQubitManager::Release(Qubit q)
+int32_t CQubitManager::GetQubitId(Qubit q)
+```
+
+---
+
+`RuntimeManagement.cpp`
+
+Implementation of the `IRuntimeDriver` interface is straightforward.
+Qubit management is delegated to the respective `QubitManager` functions, taking care to add or remove qubits from the compute register and update the state vector accordingly.
+Note that new qubits are simply appended to the end of the register:
+
+```cpp
+Qubit StateSimulator::AllocateQubit()
+{
+    Qubit q = this->qbm->Allocate();
+    this->computeRegister.push_back(q);
+    UpdateState(this->numActiveQubits++);  // |Ψ'⟩ = |Ψ⟩ ⊗ |0⟩
+    return q;
+}
+
+void StateSimulator::ReleaseQubit(Qubit q)
+{
+    UpdateState(GetQubitIdx(q), /*remove=*/true);  // ρ' = tr_i[|Ψ⟩〈Ψ|]
+    this->numActiveQubits--;
+    this->computeRegister.erase(this->computeRegister.begin() + GetQubitIdx(q));
```

Comment:
One thing to note about this implementation for qubit management is that it assumes a given qubit is never released before a qubit that was allocated after it (qubits are strictly First-In-Last-Out ordered). That is true for Q# due to enforced qubit scoping, but may not be true for QIR generally when generated from arbitrary languages.

Reply:
I think it would be worth calling this out in the description above along with the note that newly allocated qubits are appended to the end of the register.